lib/std/special/compiler_rt/clzsi2.zig

```diff
@@ -0,0 +1,150 @@
+// Ported from:
+//
+// https://github.com/llvm-mirror/compiler-rt/blob/f0745e8476f069296a7c71accedd061dce4cdf79/lib/builtins/clzsi2.c
+// https://github.com/llvm-mirror/compiler-rt/blob/f0745e8476f069296a7c71accedd061dce4cdf79/lib/builtins/arm/clzsi2.S
+const builtin = @import("builtin");
+
+// Precondition: a != 0
+fn __clzsi2_generic(a: i32) callconv(.C) i32 {
+    @setRuntimeSafety(builtin.is_test);
+
+    var x: c_uint = @bitCast(c_uint, a);
+    var t: c_int = @intCast(c_int, @boolToInt((x & 0xFFFF0000) == 0)) << 4; // if (x is small) t = 16 else 0
+    x >>= @intCast(u5, 16 - t); // x = [0 - 0xFFFF]
+
+    var r: c_int = t; // r = [0, 16]
+     // return r + clz(x)
+    t = @intCast(c_int, @boolToInt((x & 0xFF00) == 0)) << 3;
+    x >>= @intCast(u5, 8 - t); // x = [0 - 0xFF]n
+    r += t; // r = [0, 8, 16, 24]
+    // return r + clz(x)
+    t = @intCast(c_int, @boolToInt((x & 0xF0) == 0)) << 2;
+    x >>= @intCast(u5, 4 - t); // x = [0 - 0xF]
+    r += t;  // r = [0, 4, 8, 12, 16, 20, 24, 28]
+    // return r + clz(x)
+    t = @intCast(c_int, @boolToInt((x & 0xC) == 0)) << 1;
+    x >>= @intCast(u5, 2 - t); // x = [0 - 3]
+    r += t; // r = [0 - 30] and is even
+    // return r + clz(x)
+    //     switch (x)
+    //     {
+    //     case 0:
+    //         return r + 2;
+    //     case 1:
+    //         return r + 1;
+    //     case 2:
+    //     case 3:
+    //         return r;
+    //     }
+    return @intCast(i32, r) + ((2 - @intCast(i32,x)) & -@intCast(i32, @boolToInt((x & 2) == 0)));
+}
+
+fn __clzsi2_arm_clz(a: i32) callconv(.C) i32 {
+    return asm volatile(
```

Comment:
If you aim for efficiency you probably want something like

```zig
fn __clzsi2_arm_clz() callconv(.Naked) void {
    asm volatile(
        \\ clz r0, r0
        \\ bx lr
    );
    unreachable;
}
```

This way you don't have any frame setup/teardown overhead and no redundant input/output/clobbers in the code (remember that the final set of clobbered registers is the union of the in/out/clobber indications).

Reply:
I tried that but the test suite would fail on ARM platforms before you can't call a naked function from a non-naked function

Reply:
Was the problem in the test code?

Reply:
> you can't call a naked function from a non-naked function

You have to `@ptrCast` the function to a different calling convention first. If the assembly code in your naked function matches one of the calling conventions then it is safe to cast the function to that and call it.